### DISCUSSION TITLE
Can/should sqlalchemy keep pooled connections alive by pinging?

### DISCUSSION THREAD

**Discussion by amacfie-tc** [sqlalchemy/sqlalchemy#8943] [Usage Questions] [answered]
`create_engine` has the `pool_recycle` argument so that sqlalchemy knows when a connection should no longer be assumed to be connected. As an alternative, is there a way, or should there be a way, for sqlalchemy to send `SELECT 1;` periodically on connections in the pool so that they don't get disconnected?

**Comment by zzzeek** [comment, marked as answer]
yes this is called [pool_pre_ping](https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine.params.pool_pre_ping) and it's ready to go.  See https://docs.sqlalchemy.org/en/14/core/pooling.html#pool-disconnects-pessimistic

**Reply by amacfie-tc** [reply]
But pool_pre_ping doesn't prevent the connection from being disconnected, or am I wrong?

**Reply by zzzeek** [reply]
if the application is idle and the server terminates a connection, then yes, pool_pre_ping hasn't taken place at that point unless you have taken extra steps so that it does.

if you want the connection to never actually be terminated even if the application is "idle", you have to modify the application so that it's not actually "idle", by implementing a worker thread that checks out connections periodically to keep them fresh.   spawning a background thread is out of scope for SQLAlchemy which is only a library, not an application server.

**Reply by zzzeek** [reply]
I would consider though, why you might think you'd need something like that.  your application is idle, so performance is not an issue, unless you are going for very instantaneous wakeup time when a request does come in after a long idle period.  this sounds like a GUI app in that case where you would have ready access to background threads (unlike in a web application).

**Reply by amacfie-tc** [reply]
Yeah I was considering a situation where requests arrive infrequently so connections might be killed but also where we'd want to minimize delay when a request does come in since reconnecting seems to be time consuming. I'll think about the points you raised.
